index.js

```diff
@@ -96,54 +135,146 @@ original error: ${e}`);
+
+function handleSpaces(s) {
+  ret = s;
+  if (argv.nospaces) {
+    ret = s.replaceAll(' ', '_');
+  }
+  return ret;
+}
+
+
+async function buildPodcastDict(fileName, cacheFilesPath, podcastsDBData) {
+  const uuid = fileName.replace(".mp3", "");
+  const dbMeta = podcastsDBData.find((m) => m.zuuid === uuid);
+  const path = `${cacheFilesPath}/${fileName}`;
+  const exportBase = dbMeta?.zcleanedtitle // 1. from apple podcast database
+        ?? (await getMP3MetaTitle(path)) // 2. from mp3 meta data
+        ?? uuid; // 3. fallback to unreadable uuid
+  const podcastName = dbMeta?.zpodcast.replaceAll('/', '_');
+  const exportFileName = sanitize(exportBase.substr(0, fileNameMaxLength));
+
+  const ret = {
+    podcastName: handleSpaces(podcastName),
+    date: dbMeta?.date,
+    fileName,
+    path,
+    uuid,
+    exportFileName: handleSpaces(`${exportFileName}.mp3`)
+  };
+  return ret;
+}
+
+
+function filterPodcasts(podcasts, filepatterns = []) {
+  if (filepatterns.length == 0) {
+    return podcasts;
+  }
+
+  function matchesAny(s) {
+    return filepatterns.some((p) => { return s.indexOf(p) != -1 })
+  }
+
+  return podcasts.filter((p) => {
+    return matchesAny(p.exportFileName) || matchesAny(p.podcastName);
+  });
+}
+
+
+async function exportSingle(podcast, newPath) {
+  copyFileSync(podcast.path, newPath);
+  if (podcast.date && argv.updateutime) {
+    const d = new Date(podcast.date);
+    await fs.utimes(newPath, d, d);
+  }
+}
+
+
+async function exportPodcasts(podcastsDBData, filepatterns = []) {
   const cacheFilesPath = await getPodcastsCacheFilesPath();
   const podcastMP3Files = await getPodcastsCacheMP3Files(cacheFilesPath);
-  const filesWithDBData = podcastMP3Files.map((fileName) => {
-    const uuid = fileName.replace(".mp3", "");
-    const dbMeta = podcastsDBData.find((m) => m.zuuid === uuid);
-    return {
-      fileName,
-      uuid,
-      path: `${cacheFilesPath}/${fileName}`,
-      dbMeta
-    };
-  });
+  const podcasts = await Promise.all(podcastMP3Files.map((fileName) => {
+    return buildPodcastDict(fileName, cacheFilesPath, podcastsDBData);
+  }));
+  let filteredPodcasts = filterPodcasts(podcasts, filepatterns);
+
+  // Weirdly, there are some podcasts that are duplicates ...  i.e.,
+  // if you uncomment the below code, it prints some duplicate names,
+  // in my case at least.
+  /*
+    const allNames = filteredPodcasts.map(p => p.exportFileName);
+    console.log(`have ${allNames.length} names`);
+    const uniqueNames = Array.from(new Set(allNames));
+    console.log(`have ${uniqueNames.length} names`);
+    const dups = allNames.filter((e, i, a) => a.indexOf(e) !== i);
+    console.log(`duplicates! ${dups}`);
+    process.exit(0);
+  */
+  // Since this causes some strange messages to appear during output
+  // to a new output directory, delete the dups by keying on filename.
+  const temp = {}
+  // console.log('removing dups');
+  filteredPodcasts.forEach(p => temp[p.exportFileName] = p);
```

Comment:
I'm still not sure why duplication is a problem, but regarding this uniq values I think Object.fromEntries might look a bit better
```suggestion
  const temp = Object.fromEntries(filteredPodcasts.map(podcast => [podcast.exportFileName, podcast]))
```

Reply:
sure, np.

Reply:
Re "I'm still not sure why duplication is a problem" - just because it looks weird on output.  :-)  e.g., in my case:

```
$ node index.js CHURCHI
Exporting 2 of 464
344DA728-FE9D-40CC-92F6-2E27260EF79E.mp3 -> La Segunda Guerra Mundial (EPT)/Episodio 65. Churchill, Stalin, Roosevelt Y Sus Pr.mp3
2D78D27E-AEF7-48E2-BCA8-E270145E7BDD.mp3 -> La Segunda Guerra Mundial (EPT)/Episodio 65. Churchill, Stalin, Roosevelt Y Sus Pr.mp3
```

It's not a _problem_, it just looks/feels weird.

Reply:
Changed to `const uniqByFilename = Object.fromEntries(filteredPodcasts.map(p => [p.exportFileName, p]));`, thanks.  :-)